File: app/src/main/java/com/armageddon/android/flickrdroid/ui/activities/SettingsActivity.java
package com.armageddon.android.flickrdroid.ui.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import com.armageddon.android.flickrdroid.R;
import com.armageddon.android.flickrdroid.common.ActivityUtils;
import com.armageddon.android.flickrdroid.ui.fragments.SettingsFragment;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

/**
 * Settings class. Contains SettingsFragment.
 * Launch SearchActivity if user switches APP theme.
 */

public class SettingsActivity extends AppCompatActivity implements SettingsFragment.CallBacks {

    private static final String GALLERY_IN_ROW = "gallery_in_row";
    private static final String THEME_FLAG = "theme_flag";
    private boolean isThemeChanged;

    public static Intent newIntent(Context context, boolean isThemeChanged) {
        Intent intent = new Intent(context, SettingsActivity.class);
        intent.putExtra(THEME_FLAG, isThemeChanged);
        return intent;
    }

    public static Intent newIntent(Context context) {
        return new Intent(context, SettingsActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        ActivityUtils.onActivityCreateSetTheme(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);
        isThemeChanged = getIntent().getBooleanExtra(THEME_FLAG, false);

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setNavigationOnClickListener(v -> onBackPressed());

        FragmentManager fm = getSupportFragmentManager();
        Fragment fragment = fm.findFragmentById(R.id.fragment_container);

        if (fragment == null) {
            fragment = SettingsFragment.newInstance();
            fm.beginTransaction()
                    .add(R.id.fragment_container, fragment)
                    .commit();
        }
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    @Override
    public void onBackPressed() {
        if (isThemeChanged) {
           Intent intent = SearchActivity.newIntent(this,
                   null, SearchActivity.SHOW_MODE, false);
           intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
           startActivity(intent);
        } else {
            super.onBackPressed();
        }
    }


    /**
     * Sends signal to previous activities if user changed the number of rows
     */
    @Override
    public void onGalleryInRowChanged() {
        Intent data = new Intent();
        data.putExtra(GALLERY_IN_ROW,true);
        setResult(RESULT_OK,data);
    }
}

File: app/src/main/java/com/armageddon/android/flickrdroid/ui/activities/SettingsActivity.kt
package com.armageddon.android.flickrdroid.ui.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import com.armageddon.android.flickrdroid.BuildConfig
import com.armageddon.android.flickrdroid.R
import com.armageddon.android.flickrdroid.common.ActivityUtils
import com.armageddon.android.flickrdroid.common.AppPreferences
import com.armageddon.android.flickrdroid.databinding.ActivityAboutBinding
import com.armageddon.android.flickrdroid.ui.fragments.LogoFragment
import com.armageddon.android.flickrdroid.ui.fragments.SettingsFragment

private const val ABOUT_FRAGMENT = "about_fragment"

class SettingsActivity : SingleActivity(), SettingsFragment.CallBacks, LogoFragment.CallBacks {
    private var mColumnsChanged = false
    private var mInflateAboutFragment = false

    companion object {
        fun newIntent(context: Context, inflateAboutFragment: Boolean = false) : Intent {
            return Intent(context, SettingsActivity::class.java)
                .putExtra(ABOUT_FRAGMENT, inflateAboutFragment)
        }
    }

    override fun getFragment(): Fragment {
       return when (mInflateAboutFragment) {
           true -> LogoFragment.newInstance(true)
           false -> SettingsFragment.newInstance()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        mInflateAboutFragment = intent.getBooleanExtra(ABOUT_FRAGMENT, false)
        super.onCreate(savedInstanceState)
        if (mInflateAboutFragment) {
            val bindingAbout = ActivityAboutBinding.inflate(layoutInflater)
            setContentView(bindingAbout.root)
            setSupportActionBar(bindingAbout.toolbar)
            val version = BuildConfig.VERSION_NAME

            bindingAbout.apply {
                versionView.text = getString(R.string.app_version, version)
                listOfChanges.setOnClickListener { showListOfChangesDialog() }
            }
        }


    }


    override fun onBackPressed() {
        val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (fragment is LogoFragment) {
            setContentView(R.layout.activity_single_fragment)
        }

        when (ActivityUtils.isThemeChanged || mColumnsChanged) {
            true -> {
                if (AppPreferences.getOauthToken(this).isNullOrBlank())
                    startActivity(TopPhotoActivity.newIntent(this))
                else
                    startActivity(TopPhotoActivity.newIntent(this))
                ActivityUtils.isThemeChanged = false
            }
            false -> {
                super.onBackPressed()
            }
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }


    override fun onThemeChanged() {
        val intent = newIntent(this)
        startActivity(intent)
        finish()
        overridePendingTransition(0, 0)
    }

    override fun onColumnsChanged() {
        mColumnsChanged = true
    }

    override fun onAboutClick() {
        val intent = newIntent(this, true)
        startActivity(intent)
        overridePendingTransition(0, 0)
    }

    override fun onLogoAnimationEnd() {

    }

    private fun showListOfChangesDialog () {
        val builder = AlertDialog.Builder(this, R.style.DialogAllCapsFalseStyle).apply {
            setMessage(getString(R.string.text_of_changes))
            setPositiveButton(R.string.ok) { _, _ -> }
        }
        val alert = builder.create()
        alert.show()
    }
}
